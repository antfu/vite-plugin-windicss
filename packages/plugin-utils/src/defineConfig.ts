import { FullConfig } from 'windicss/types/interfaces'

export function defineConfig(config: FullConfig) {
  return config
}
