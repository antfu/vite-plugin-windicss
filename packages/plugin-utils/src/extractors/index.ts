export * from './default'
export * from './helper'
export * from './pug'
