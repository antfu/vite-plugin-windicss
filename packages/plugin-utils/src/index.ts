export * from './constants'
export * from './defineConfig'
export * from './createUtils'
export * from './options'
export * from './transformers'
export * from './utils'
export * from './resolveOptions'
