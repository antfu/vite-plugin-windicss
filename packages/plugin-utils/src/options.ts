/* eslint-disable no-use-before-define */

import type { Extractor, FullConfig as WindiCssOptions } from 'windicss/types/interfaces'
import { TransformerFunction } from './transformers'
import { TagNames } from './constants'

export { WindiCssOptions }

export interface UserOptions {
  /**
   * Options for windicss/tailwindcss.
   * Also accepts string as config file path.
   *
   * @default auto searching for `windi.config.ts` / `tailwind.config.js`
   */
  config?: WindiCssOptions | string

  /**
   * Safe class names to be always included.
   *
   * @deprecated define this field in the windicss.config.ts instead
   */
  safelist?: string | (string | string[])[]

  /**
   * Class names to be always excluded.
   *
   * @deprecated define this field in the windicss.config.ts instead
   */
  blocklist?: string | (string | string[])[]

  /**
   * Enabled windicss preflight (a.k.a TailwindCSS style reset)
   *
   * @deprecated define this field in the windicss.config.ts instead
   * @default true
   */
  preflight?: boolean | {
    /**
     * Enable all the preflight regardless the template
     *
     * @deprecated
     */
    enableAll?: boolean

    /**
     * Enable all the preflight regardless the template
     *
     * @deprecated define this field in the windicss.config.ts instead
     */
    includeAll?: boolean

    /**
     * Safelist to always included
     *
     * @deprecated define this field in the windicss.config.ts instead
     */
    safelist?: string | (string | string[])[]

    /**
     * Blocklist to always excluded
     *
     * @deprecated define this field in the windicss.config.ts instead
     */
    blocklist?: string | (string | string[])[]

    /**
      * Alias for resolving preflight
      */
    alias?: Record<string, TagNames>

    /**
     * @default true
     * @deprecated define this field in the windicss.config.ts instead
     */
    includeBase?: boolean

    /**
     * @default true
     * @deprecated define this field in the windicss.config.ts instead
     */
    includeGlobal?: boolean

    /**
     * @default true
     * @deprecated define this field in the windicss.config.ts instead
     */
    includePlugin?: boolean
  }

  /**
    * File paths will be resolved against this directory.
    *
    * @default process.cwd
    * @internal
    */
  root?: string

  /**
   * Scan the files and extract the usage
   *
   * @default true
   */
  scan?: boolean | {
    /**
     * Auto scan on startup
     *
     * @default true
     */
    runOnStartup?: boolean

    /**
     * Directories to search for classnames
     *
     * @default 'src'
     */
    dirs?: string | string[]

    /**
     * File extension to search for classnames
     *
     * @default 'html', 'vue', 'md', 'mdx', 'pug', 'jsx', 'tsx', 'svelte'
     */
    fileExtensions?: string | string[]

    /**
     * Exclude globs
     *
     * @default []
     */
    exclude?: string | string[]

    /**
     * Include globs
     *
     * @default []
     */
    include?: string | string[]

    /**
     * Transformers to apply before doing extraction
     *
     * @default []
     */
    transformers?: TransformerFunction[]
  }

  /**
   * Transform CSS for `@apply` directive
   *
   * @default true
   */
  transformCSS?: boolean | 'pre' | 'post'

  /**
   * Transform groups like `hover:(bg-gray-100 font-medium)`
   *
   * @default true
   */
  transformGroups?: boolean

  /**
   * Sort the genrate utilities
   *
   * @default true
   */
  sortUtilities?: boolean

  /**
   * Callback before classes css generated
   */
  onBeforeGenerate?: (ctx: { classesPending: Set<string>; tagsPending: Set<string> }) => void

  /**
   * Callback when classes and/or tags are generated/changed
   */
  onGenerated?: (ctx: { classes: Set<string>; tags: Set<string> }) => void

  /**
   * Callback when the options are resolved. These are the plugin options and contain the windi config
   */
  onOptionsResolved?: (options: ResolvedOptions) => ResolvedOptions | void | Promise<ResolvedOptions | void>

  /**
   * Callback when the windi config is resolved. Not to be confused with the options which are the top level way to
   * configure the util package
   */
  onConfigResolved?: (config: WindiCssOptions, configFilePath?: string) => WindiCssOptions | void | Promise<WindiCssOptions | void>
}

export interface WindiPluginUtilsOptions {
  /**
   * Name for debug
   *
   * @default 'windi-plugin-utils'
   * @internal
   */
  name?: string

  /**
    * CWD
    *
    * @default process.cwd
    * @internal
    */
  root?: string

  /**
   * Use sucrase/register to load configs in ESM/TypeScript
   *
   * @default true
   */
  enableSucrase?: boolean
}

export interface ResolvedOptions {
  config: WindiCssOptions
  configFilePath: string | undefined

  enableScan: boolean
  enablePreflight: boolean
  transformCSS: boolean | 'pre' | 'auto' | 'post'
  transformGroups: boolean

  scanOptions: {
    fileExtensions: string[]
    dirs: string[]
    exclude: string[]
    include: string[]
    runOnStartup: boolean
    transformers: TransformerFunction[]
    extractors: Extractor[]
  }

  preflightOptions: {
    includeBase: boolean
    includeGlobal: boolean
    includePlugin: boolean
    includeAll: boolean
    /**
     * @deprecated use includeAll
     */
    enableAll: boolean
    safelist: Set<string>
    blocklist: Set<string>
    alias: Record<string, string>
  }

  root: string
  sortUtilities: boolean
  safelist: Set<string>
  blocklist: Set<string>
  onBeforeGenerate: UserOptions['onBeforeGenerate']
  onGenerated: UserOptions['onGenerated']
  onConfigResolved: UserOptions['onConfigResolved']
  onOptionsResolved: UserOptions['onOptionsResolved']
}
