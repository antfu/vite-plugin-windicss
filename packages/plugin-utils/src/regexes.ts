export const regexQuotedString = /(["'`])((?:\\\1|(?:(?!\1)|\n).)*?)\1/mg
export const regexClassCheck = /^[a-z\d.+-]+[\w:/\\.$-]*\w$/
export const regexHtmlTag = /<(\w[\w-]*)/g
export const regexClassSplitter = /[\s'"`{}]/g
export const regexClassGroup = /(\w[\w:_/-]*?):\(([\w\s/-]*?)\)/gm
