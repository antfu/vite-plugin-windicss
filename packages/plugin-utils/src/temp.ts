import { Config } from 'windicss/types/interfaces'

export interface ExtractorResultDetailed {
  attributes?: {
    names: string[]
    values: string[]
  }
  classes?: string[]
  ids?: string[]
  tags?: string[]
  undetermined?: string[]
}

export interface Extractor {
  extractor: (content: string) => ExtractorResultDetailed | Promise<ExtractorResultDetailed>
  extensions: string[]
}

export interface ExtractOptions {
  /**
   * Globs of files to be included from extractions
   */
  include?: string[]
  /**
   * Globs of files to be excluded from extractions
   *
   * @default ['node_modules', '.git']
   */
  exclude?: string[]
  /**
   * Extractors to handle different file types.
   * Compatible with PurgeCSS
   */
  extractors?: Extractor[]
}

export interface PreflightOptions {
  /**
   * Include all the preflight regardless the template
   *
   * @default false
   */
  includeAll?: boolean

  /**
   * Safelist of preflight that will always be included in the generated CSS
   */
  safelist?: string | (string | string[])[]

  /**
   * Blocklist of preflight that will always be excluded in the generated CSS
   */
  blocklist?: string | (string | string[])[]

  /**
   * Alias for resolving preflight
   */
  alias?: Record<string, string>

  /**
   * @default true
   */
  includeBase?: boolean

  /**
   * @default true
   */
  includeGlobal?: boolean

  /**
   * @default true
   */
  includePlugin?: boolean
}

export interface FullConfig extends Config {
  /**
   * Safelist of utilities that will always be included in the generated CSS
   */
  safelist?: string | (string | string[])[]
  /**
   * Blocklist of utilities that will always be excluded in the generated CSS
   */
  blocklist?: string | (string | string[])[]
  /**
   * Extractions options
   */
  extract?: ExtractOptions
  /**
   * Preflight options
   * Set `false` to disable preflight
   *
   * @default true
   */
  preflight?: PreflightOptions | boolean
}
