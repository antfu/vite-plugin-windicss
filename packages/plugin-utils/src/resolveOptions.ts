import { existsSync } from 'fs'
import { resolve, posix } from 'path'
import _debug from 'debug'
import { defaultAlias, configureFiles } from './constants'
import { Arrayable, kebabCase, mergeArrays, toArray } from './utils'
import { UserOptions, ResolvedOptions, WindiCssOptions, WindiPluginUtilsOptions } from './options'
import { PugTransformer, TransformerFunction } from './transformers'

export function isResolvedOptions(options: UserOptions | ResolvedOptions): options is ResolvedOptions {
  // @ts-expect-error internal flag
  return options.__windi_resolved
}

export function getDefaultTransformers() {
  const transformers: TransformerFunction[] = []

  // auto detect pug
  try {
    require.resolve('pug')
    transformers.push(
      PugTransformer(),
    )
  }
  catch (e) {}

  return transformers
}

function buildGlobs(dirs: Arrayable<string>, fileExtensions: Arrayable<string>) {
  dirs = toArray(dirs)
  const exts = toArray(fileExtensions)

  const globs = exts.length
    ? dirs.map(i =>
      posix.join(
        i,
        exts.length > 1
          ? `**/*.{${exts.join(',')}}`
          : `**/*.${exts[0]}`,
      ),
    )
    : []

  return globs
}

export async function resolveOptions(
  options: UserOptions | ResolvedOptions = {},
  utilsOptions: WindiPluginUtilsOptions = {},
  loadConfigFile = false,
): Promise<ResolvedOptions> {
  if (isResolvedOptions(options))
    return options

  const { resolved: config, configFilePath } = loadConfigFile
    ? await loadConfiguration(options, utilsOptions)
    : { resolved: {} as WindiCssOptions, configFilePath: {} }

  const {
    root = utilsOptions.root || process.cwd(),
    scan = true,
    preflight = true,
    transformCSS = true,
    transformGroups = true,
    sortUtilities = true,
  } = options

  const preflightOptions = Object.assign(
    {
      includeBase: true,
      includeGlobal: true,
      includePlugin: true,
      enableAll: false,
      includeAll: false,
      safelist: [],
      blocklist: [],
      alias: {},
    },
    typeof config.preflight === 'boolean' ? {} : config.preflight,
    typeof preflight === 'boolean' ? {} : preflight,
  ) as unknown as ResolvedOptions['preflightOptions']

  // backward compatibility
  preflightOptions.includeAll = preflightOptions.includeAll || preflightOptions.enableAll

  const scanOptions = Object.assign(
    {
      fileExtensions: ['html', 'vue', 'md', 'mdx', 'pug', 'jsx', 'tsx', 'svelte'],
      dirs: ['src'],
      exclude: ['node_modules', '.git'],
      include: ['index.html'],
      runOnStartup: true,
      transformers: getDefaultTransformers(),
    },
    typeof scan === 'boolean' ? {} : scan,
  )

  scanOptions.exclude = mergeArrays(config.extract?.exclude, scanOptions.exclude)
  scanOptions.include = mergeArrays(config.extract?.include, scanOptions.include, buildGlobs(scanOptions.dirs, scanOptions.fileExtensions))

  const safelist = new Set(mergeArrays(config.safelist, options.safelist).flatMap(i => i.split(' ')))
  const blocklist = new Set(mergeArrays(config.blocklist, options.blocklist).flatMap(i => i.split(' ')))

  // preflightOptions from config file
  const configPreflightOptions = typeof config.preflight === 'boolean' ? {} : config.preflight || {}

  preflightOptions.safelist = new Set<string>(
    mergeArrays(
      configPreflightOptions?.safelist,
      Array.from(preflightOptions.safelist),
    )
      .flatMap(i => i.split(' '))
      .map((i) => {
        // selector
        const match = i.match(/^\[(.*?)\]$/)?.[1]
        if (match)
          return `div ${match}`
        return i
      }))

  preflightOptions.blocklist = new Set<string>(
    mergeArrays(
      configPreflightOptions?.blocklist,
      Array.from(preflightOptions.blocklist),
    )
      .flatMap(i => i.split(' ')),
  )

  preflightOptions.alias = Object.fromEntries(
    Object
      .entries({
        ...defaultAlias,
        ...configPreflightOptions.alias,
        ...preflightOptions.alias,
      })
      .filter(([k, v]) => [kebabCase(k), v]),
  )

  let resolvedOptions = {
    ...options,
    root,
    config,
    configFilePath,
    enableScan: Boolean(scan),
    scanOptions,
    enablePreflight: config.preflight !== false && Boolean(preflight),
    preflightOptions,
    transformCSS,
    transformGroups,
    sortUtilities,
    safelist,
    blocklist,
    __windi_resolved: true,
  } as ResolvedOptions

  // allow the resolved options to be overwritten
  const modifiedOptions = await resolvedOptions.onOptionsResolved?.(resolvedOptions)
  if (modifiedOptions != null && modifiedOptions !== resolvedOptions)
    resolvedOptions = Object.assign(resolvedOptions, modifiedOptions)

  return resolvedOptions
}

export async function loadConfiguration(options: UserOptions, utilsOptions: WindiPluginUtilsOptions) {
  let resolved: WindiCssOptions = {}
  let configFilePath: string | undefined
  let error: Error | undefined

  const {
    name = 'windicss-plugin-utils',
    enabledTypeScriptConfig = true,
  } = utilsOptions

  const {
    config,
    root = utilsOptions.root || process.cwd(),
  } = options

  const debug = _debug(`${name}:config`)

  if (typeof config === 'string' || !config) {
    if (!config) {
      for (const name of configureFiles) {
        const tryPath = resolve(root, name)
        if (existsSync(tryPath)) {
          configFilePath = tryPath
          break
        }
      }
    }
    else {
      configFilePath = resolve(root, config)
      if (!existsSync(configFilePath)) {
        console.warn(`[${name}] config file "${config}" not found, ignored`)
        configFilePath = undefined
      }
    }

    if (configFilePath) {
      try {
        debug('loading from ', configFilePath)

        if (enabledTypeScriptConfig && configFilePath.endsWith('.ts'))
          require('sucrase/register/ts')

        delete require.cache[require.resolve(configFilePath)]
        resolved = require(configFilePath)
        if (resolved.default)
          resolved = resolved.default
      }
      catch (e) {
        console.error(`[${name}] failed to load config "${configFilePath}"`)
        console.error(`[${name}] ${e.toString()}`)
        error = e
        configFilePath = undefined
        resolved = {}
      }
    }
  }
  else {
    resolved = config
  }

  // allow to hook into resolved config
  const modifiedConfigs = await options.onConfigResolved?.(resolved, configFilePath)
  if (modifiedConfigs != null)
    resolved = modifiedConfigs

  debug(resolved)

  return {
    error,
    resolved,
    configFilePath,
  }
}
