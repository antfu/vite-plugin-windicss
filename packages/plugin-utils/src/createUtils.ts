import { promises as fs, existsSync } from 'fs'
import { join, resolve } from 'path'
import WindiCssProcessor from 'windicss'
import { StyleSheet } from 'windicss/utils/style'
import { CSSParser } from 'windicss/utils/parser'
import fg from 'fast-glob'
import _debug from 'debug'
import micromatch from 'micromatch'
import { preflightTags, htmlTags, configureFiles, tagsEnableAttrs } from './constants'
import { regexQuotedString, regexClassSplitter, regexClassCheck, regexHtmlTag } from './regexes'
import { resolveOptions, WindiCssOptions, WindiPluginUtilsOptions, UserOptions, ResolvedOptions } from './options'
import { kebabCase, include, exclude, slash, transformGroups, transformGroupsWithSourcemap } from './utils'

export function createUtils(
  userOptions: UserOptions | ResolvedOptions = {},
  utilsOptions: WindiPluginUtilsOptions = {},
) {
  const options = resolveOptions(userOptions)

  const {
    config,
    scan: enabledScan,
    scanOptions,
    transformGroups: enableGroupsTransform,
    transformCSS: enableCssTransform,
    preflight: enablePreflight,
    preflightOptions,
    root = utilsOptions.root || process.cwd(),
    sortUtilities,
    safelist,
    blocklist,
  } = options

  const {
    name = 'windicss-plugin-utils',
    enabledTypeScriptConfig = true,
  } = utilsOptions

  const debug = {
    config: _debug(`${name}:config`),
    debug: _debug(`${name}:debug`),
    compile: _debug(`${name}:compile`),
    glob: _debug(`${name}:scan:glob`),
    scanTransform: _debug(`${name}:scan:transform`),
    detectClass: _debug(`${name}:detect:class`),
    detectTag: _debug(`${name}:detect:tag`),
    detectAttr: _debug(`${name}:detect:attr`),
  }

  let processor: WindiCssProcessor
  let configFilePath: string | undefined

  const globs = getGlobs()
  const excludeGlobs = getExcludeGlob()
  const files: string[] = []

  const regexId = new RegExp(`\\.(?:${scanOptions.fileExtensions.join('|')})$`, 'i')

  const configSafelist = new Set<string>()

  const classesGenerated = new Set<string>()
  const classesPending = new Set<string>()
  const tagsGenerated = new Set<string>()
  const tagsPending = new Set<string>()
  const attrsGenerated = new Set<string>()
  const attrsPending = new Set<string>()
  const tagsAvailable = new Set<string>()

  function loadConfiguration() {
    let resolved: WindiCssOptions = {}
    if (typeof config === 'string' || !config) {
      let path = ''
      if (!config) {
        for (const name of configureFiles) {
          const tryPath = resolve(root, name)
          if (existsSync(tryPath)) {
            path = tryPath
            break
          }
        }
      }
      else {
        path = resolve(root, config)
        if (!existsSync(path)) {
          console.warn(`[${name}] config file "${config}" not found, ignored`)
          path = ''
        }
      }

      if (path) {
        try {
          debug.config('loading from ', path)

          if (enabledTypeScriptConfig && path.endsWith('.ts')) {
            // eslint-disable-next-line @typescript-eslint/no-var-requires
            require('esbuild-register')
          }

          delete require.cache[require.resolve(path)]
          // eslint-disable-next-line @typescript-eslint/no-var-requires
          resolved = require(path)
          // @ts-expect-error esm resolve
          if (resolved.default) resolved = resolved.default

          configFilePath = path
          configSafelist.clear()
          // @ts-expect-error
          include(configSafelist, resolved?.purge?.options?.safelist || resolved?.purge?.options?.whitelist || [])
        }
        catch (e) {
          console.error(`[${name}] failed to load config "${path}"`)
          console.error(`[${name}] ${e.toString()}`)
          setTimeout(() => process.exit(1))
        }
      }
    }
    else {
      resolved = config
    }

    debug.config(JSON.stringify(resolved, null, 2))
    return resolved
  }

  function initWindicss() {
    return new WindiCssProcessor(loadConfiguration())
  }

  function getGlobs() {
    const { dirs, fileExtensions, include } = scanOptions
    const globs = fileExtensions.length
      ? dirs.map(i => slash(
        join(
          i,
          fileExtensions.length > 1
            ? `**/*.{${fileExtensions.join(',')}}`
            : `**/*.${fileExtensions[0]}`,
        ),
      ))
      : []

    globs.unshift('index.html')
    globs.unshift(...include)

    debug.glob('globs', globs)
    return globs
  }

  function getExcludeGlob() {
    return ['node_modules', '.git', ...scanOptions.exclude]
  }

  async function getFiles() {
    const files = await fg(
      globs,
      {
        cwd: root,
        ignore: excludeGlobs,
        onlyFiles: true,
        absolute: true,
      },
    )

    files.sort()

    debug.glob('files', files)

    return files
  }

  let scanned = false
  let _searching: Promise<void> | null

  async function scan() {
    if (!_searching) {
      _searching = (async() => {
        files.push(...await getFiles())

        const contents = await Promise.all(
          files
            .filter(id => isDetectTarget(id))
            .map(async id => [await fs.readFile(id, 'utf-8'), id]),
        )

        for (const [content, id] of contents)
          extractFile(content, id, true)

        scanned = true
      })()
    }

    return _searching
  }

  function isExcluded(id: string) {
    return id.match(/\b(?:node_modules|.git)\b/)
      || micromatch.isMatch(slash(id), excludeGlobs)
  }

  function isDetectTarget(id: string) {
    if (files.some(file => id.startsWith(file)))
      return true
    return id.match(regexId) && !isExcluded(id)
  }

  function isScanTarget(id: string) {
    return enabledScan
      ? files.some(file => id.startsWith(file))
      : isDetectTarget(id)
  }

  function isCssTransformTarget(id: string) {
    if (id.match(/\.(?:postcss|scss|sass|css|stylus|less)(?:$|\?)/i) && !isExcluded(id))
      return true
    return false
  }

  function extractFile(code: string, id?: string, applyGroupTransform = true) {
    if (applyGroupTransform) {
      if (enableGroupsTransform)
        code = transformGroups(code)
    }

    if (id) {
      debug.scanTransform(id)
      for (const trans of scanOptions.transformers) {
        const result = trans(code, id)
        if (result != null)
          code = result
      }
    }

    let changed = false
    // classes
    Array.from(code.matchAll(regexQuotedString))
      .flatMap(m => (m[2] || '').split(regexClassSplitter))
      .filter(i => i.match(regexClassCheck))
      .forEach((i) => {
        if (!i || classesGenerated.has(i) || classesPending.has(i) || blocklist.has(i))
          return
        classesPending.add(i)
        changed = true
      })

    if (enablePreflight || !preflightOptions.enableAll) {
      // preflight
      Array.from(code.matchAll(regexHtmlTag))
        .forEach(([full, tag]) => {
          if (!tagsAvailable.has(tag))
            tag = preflightOptions.alias[kebabCase(tag)]
          if (preflightOptions.blocklist.has(tag))
            return
          if (tagsAvailable.has(tag) && !tagsPending.has(tag)) {
            tagsPending.add(tag)
            tagsAvailable.delete(tag)
            changed = true
          }
          if (tagsEnableAttrs[tag] && !attrsPending.has(full)) {
            attrsPending.add(full)
            changed = true
          }
        })
    }

    if (changed) {
      debug.detectClass(classesPending)
      debug.detectTag(tagsPending)
      debug.detectAttr(attrsPending)
    }

    return changed
  }

  function transformCSS(css: string) {
    if (!enableCssTransform)
      return css
    const style = new CSSParser(css, processor).parse()
    return style.build()
  }

  let style: StyleSheet = new StyleSheet()
  let _cssCache: string | undefined

  async function generateCSS() {
    if (enabledScan && scanOptions.runOnStartup)
      await scan()

    options.onBeforeGenerate?.({
      classesPending,
      tagsPending,
    })

    let changed = false

    if (classesPending.size) {
      const result = processor.interpret(Array.from(classesPending).join(' '))
      if (result.success.length) {
        include(classesGenerated, result.success)
        classesPending.clear()
        debug.compile(`compiled ${result.success.length} classes`)
        debug.compile(result.success)

        style = style.extend(result.styleSheet)
        changed = true
      }
    }

    if (enablePreflight) {
      if (preflightOptions.enableAll || tagsPending.size) {
        const preflightStyle = processor.preflight(
          preflightOptions.enableAll
            ? undefined
            : [
              ...Array.from(tagsPending).map(i => `<${i}/>`),
              ...Array.from(attrsPending),
            ].join(' '),
          preflightOptions.includeBase,
          preflightOptions.includeGlobal,
          preflightOptions.includePlugin,
        )
        style = style.extend(preflightStyle, true)
        include(tagsGenerated, tagsPending)
        include(attrsGenerated, attrsPending)
        tagsPending.clear()
        attrsPending.clear()
        changed = true
      }
    }

    if (changed || !_cssCache) {
      if (sortUtilities)
        style.sort()

      _cssCache = style.build()

      options.onGenerated?.({
        classes: classesGenerated,
        tags: tagsGenerated,
      })
    }
    return _cssCache
  }

  function clearCache() {
    style = new StyleSheet()
    _cssCache = undefined

    include(classesPending, configSafelist)
    include(classesPending, safelist)
    include(classesPending, classesGenerated)

    include(tagsPending, tagsGenerated)
    include(tagsPending, preflightTags)
    include(tagsPending, preflightOptions.safelist)
    include(tagsAvailable, htmlTags as any as string[])

    exclude(tagsAvailable, preflightTags)
    exclude(tagsAvailable, preflightOptions.safelist)

    classesGenerated.clear()
    tagsGenerated.clear()
    attrsGenerated.clear()
    attrsPending.clear()
  }

  function init() {
    processor = initWindicss()
    clearCache()
  }

  return {
    options,
    init,
    extractFile,
    generateCSS,
    clearCache,
    transformCSS,
    transformGroups,
    transformGroupsWithSourcemap,
    isDetectTarget,
    isScanTarget,
    isCssTransformTarget,
    isExcluded,
    scan,

    files,
    globs,

    classesGenerated,
    classesPending,
    tagsGenerated,
    tagsPending,
    tagsAvailable,

    get scanned() {
      return scanned
    },
    get configFilePath() {
      return configFilePath
    },
    get hasPending() {
      return Boolean(tagsPending.size || classesPending.size)
    },
  }
}

export type WindiPluginUtils = ReturnType<typeof createUtils>
