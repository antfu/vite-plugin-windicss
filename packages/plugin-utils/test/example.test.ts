import { resolve } from 'path'
import { createUtils } from '../src'

describe('example', () => {
  it('should work', async() => {
    const utils = createUtils({}, {
      root: resolve(__dirname, '../../../examples/vue'),
    })
    await utils.init()
    expect(await utils.generateCSS()).toMatchSnapshot('generated-css')

    // layers
    expect(await utils.generateCSS('base')).toMatchSnapshot('generated-css-base')
    expect(await utils.generateCSS('components')).toMatchSnapshot('generated-css-components')
    expect(await utils.generateCSS('utilities')).toMatchSnapshot('generated-css-utilities')

    expect(utils.classesGenerated).toMatchSnapshot('classes')
    expect(utils.tagsGenerated).toMatchSnapshot('tags')
  })

  it('should allow access to generated tags and classes by a callback', async() => {
    let classes, tags, classesPending, tagsPending
    const utils = createUtils({
      onBeforeGenerate(ctx) {
        classesPending = new Set(ctx.classesPending)
        tagsPending = new Set(ctx.tagsPending)
      },
      onGenerated(ctx) {
        classes = ctx.classes
        tags = ctx.tags
      },
    }, {
      root: resolve(__dirname, '../../../examples/vue'),
    })
    await utils.init()
    await utils.generateCSS()
    expect(classesPending).toMatchSnapshot('classesPending')
    expect(tagsPending).toMatchSnapshot('tagsPending')
    expect(classes).toMatchSnapshot('classes')
    expect(tags).toMatchSnapshot('tags')
  })
})
