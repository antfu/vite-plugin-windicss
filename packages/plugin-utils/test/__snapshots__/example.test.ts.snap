// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`example should allow access to generated tags and classes by a callback: classes 1`] = `
Set {
  "select-none",
  "p-5",
  "px-5",
  "p-1",
  "text-white",
  "hover:text-green-100",
  "hover:rounded-full",
  "bg-teal-100",
  "hover:w-1/3",
  "btn",
  "bg-black",
  "bg-blue-200",
  "bg-red-200",
  "!bg-blue-200",
  "pl-100px",
  "bg-yellow-200",
  "mt-2",
}
`;

exports[`example should allow access to generated tags and classes by a callback: classesPending 1`] = `
Set {
  "select-none",
  "en",
  "viewport",
  "app",
  "p-5",
  "module",
  "ts",
  "./Foo.vue",
  "px-5",
  "p-1",
  "text-white",
  "hover:text-green-100",
  "hover:rounded-full",
  "bg-teal-100",
  "hover:w-1/3",
  "true",
  "btn",
  "bar",
  "bg-black",
  "bg-blue-200",
  "bg-red-200",
  "!bg-blue-200",
  "postcss",
  "scss",
  "stylus",
  "pl-100px",
  "bg-yellow-200",
  "mt-2",
}
`;

exports[`example should allow access to generated tags and classes by a callback: tags 1`] = `
Set {
  "html",
  "body",
  "div",
  "head",
  "meta",
  "title",
  "script",
  "nav",
  "h2",
  "a",
  "style",
}
`;

exports[`example should allow access to generated tags and classes by a callback: tagsPending 1`] = `
Set {
  "html",
  "body",
  "div",
  "head",
  "meta",
  "title",
  "script",
  "nav",
  "h2",
  "a",
  "style",
}
`;

exports[`example should work: classes 1`] = `
Set {
  "select-none",
  "p-5",
  "px-5",
  "p-1",
  "text-white",
  "hover:text-green-100",
  "hover:rounded-full",
  "bg-teal-100",
  "hover:w-1/3",
  "btn",
  "bg-black",
  "bg-blue-200",
  "bg-red-200",
  "!bg-blue-200",
  "pl-100px",
  "bg-yellow-200",
  "mt-2",
}
`;

exports[`example should work: generated-css 1`] = `
"*, ::before, ::after {
  -webkit-box-sizing: border-box;
  box-sizing: border-box;
  border-width: 0;
  border-style: solid;
  border-color: #e5e7eb;
}
* {
  --tw-ring-inset: var(--tw-empty,/*!*/ /*!*/);
  --tw-ring-offset-width: 0px;
  --tw-ring-offset-color: #fff;
  --tw-ring-color: rgba(59, 130, 246, 0.5);
  --tw-ring-offset-shadow: 0 0 #0000;
  --tw-ring-shadow: 0 0 #0000;
  --tw-shadow: 0 0 #0000;
}
:root {
  -moz-tab-size: 4;
  -o-tab-size: 4;
  tab-size: 4;
}
:-moz-focusring {
  outline: 1px dotted ButtonText;
}
:-moz-ui-invalid {
  box-shadow: none;
}
::moz-focus-inner {
  border-style: none;
  padding: 0;
}
::-webkit-inner-spin-button, ::-webkit-outer-spin-button {
  height: auto;
}
::-webkit-search-decoration {
  -webkit-appearance: none;
}
::-webkit-file-upload-button {
  -webkit-appearance: button;
  font: inherit;
}
[type='search'] {
  -webkit-appearance: textfield;
  outline-offset: -2px;
}
abbr[title] {
  -webkit-text-decoration: underline dotted;
  text-decoration: underline dotted;
}
body {
  margin: 0;
  font-family: inherit;
  line-height: inherit;
}
html {
  -webkit-text-size-adjust: 100%;
  font-family: ui-sans-serif,system-ui,-apple-system,BlinkMacSystemFont,\\"Segoe UI\\",Roboto,\\"Helvetica Neue\\",Arial,\\"Noto Sans\\",sans-serif,\\"Apple Color Emoji\\",\\"Segoe UI Emoji\\",\\"Segoe UI Symbol\\",\\"Noto Color Emoji\\";
  line-height: 1.5;
}
a {
  color: inherit;
  text-decoration: inherit;
}
h2 {
  margin: 0;
  font-size: inherit;
  font-weight: inherit;
}
.btn {
  border-radius: 0.25rem;
  border-width: 1px;
  --tw-border-opacity: 1;
  border-color: rgba(209, 213, 219, var(--tw-border-opacity));
  --tw-text-opacity: 1;
  color: rgba(75, 85, 99, var(--tw-text-opacity));
  padding-left: 1rem;
  padding-right: 1rem;
  padding-top: 0.5rem;
  padding-bottom: 0.5rem;
  margin: 0.5rem;
  display: inline-block;
}
.btn:hover {
  --tw-shadow: 0 1px 3px 0 rgba(0, 0, 0, 0.1), 0 1px 2px 0 rgba(0, 0, 0, 0.06);
  -webkit-box-shadow: var(--tw-ring-offset-shadow, 0 0 #0000), var(--tw-ring-shadow, 0 0 #0000), var(--tw-shadow);
  box-shadow: var(--tw-ring-offset-shadow, 0 0 #0000), var(--tw-ring-shadow, 0 0 #0000), var(--tw-shadow);
}
.bg-teal-100 {
  --tw-bg-opacity: 1;
  background-color: rgba(0, 153, 102, var(--tw-bg-opacity));
}
.bg-black {
  --tw-bg-opacity: 1;
  background-color: rgba(0, 0, 0, var(--tw-bg-opacity));
}
.bg-blue-200 {
  --tw-bg-opacity: 1;
  background-color: rgba(191, 219, 254, var(--tw-bg-opacity));
}
.bg-red-200 {
  --tw-bg-opacity: 1;
  background-color: rgba(254, 202, 202, var(--tw-bg-opacity));
}
.\\\\!bg-blue-200 {
  --tw-bg-opacity: 1 !important;
  background-color: rgba(191, 219, 254, var(--tw-bg-opacity)) !important;
}
.bg-yellow-200 {
  --tw-bg-opacity: 1;
  background-color: rgba(253, 230, 138, var(--tw-bg-opacity));
}
.hover\\\\:rounded-full:hover {
  border-radius: 9999px;
}
.mt-2 {
  margin-top: 0.5rem;
}
.p-5 {
  padding: 1.25rem;
}
.p-1 {
  padding: 0.25rem;
}
.px-5 {
  padding-left: 1.25rem;
  padding-right: 1.25rem;
}
.pl-100px {
  padding-left: 100px;
}
.text-white {
  --tw-text-opacity: 1;
  color: rgba(255, 255, 255, var(--tw-text-opacity));
}
.hover\\\\:text-green-100:hover {
  --tw-text-opacity: 1;
  color: rgba(209, 250, 229, var(--tw-text-opacity));
}
.select-none {
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}
.hover\\\\:w-1\\\\/3:hover {
  width: 33.333333%;
}"
`;

exports[`example should work: tags 1`] = `
Set {
  "html",
  "body",
  "div",
  "head",
  "meta",
  "title",
  "script",
  "nav",
  "h2",
  "a",
  "style",
}
`;
